cpp/src/arrow/compute/exec/hash_join_node.cc

```diff
@@ -43,32 +43,49 @@ bool HashJoinSchema::IsTypeSupported(const DataType& type) {
   return is_fixed_width(id) || is_binary_like(id) || is_large_binary_like(id);
 }
 
-Result<std::vector<FieldRef>> HashJoinSchema::VectorDiff(const Schema& schema,
-                                                         const std::vector<FieldRef>& a,
-                                                         const std::vector<FieldRef>& b) {
-  std::unordered_set<int> b_paths;
-  for (size_t i = 0; i < b.size(); ++i) {
-    ARROW_ASSIGN_OR_RAISE(auto match, b[i].FindOne(schema));
-    b_paths.insert(match[0]);
+Result<std::vector<FieldRef>> HashJoinSchema::ComputePayload(
+    const Schema& schema, const std::vector<FieldRef>& output,
+    const std::vector<FieldRef>& filter, const std::vector<FieldRef>& keys) {
+  // payload = (output + filter) - keys, with no duplicates
+  std::unordered_set<int> payload_fields;
+  for (auto ref : output) {
+    ARROW_ASSIGN_OR_RAISE(auto match, ref.FindOne(schema));
+    payload_fields.insert(match[0]);
   }
 
-  std::vector<FieldRef> result;
+  for (auto ref : filter) {
+    ARROW_ASSIGN_OR_RAISE(auto match, ref.FindOne(schema));
+    payload_fields.insert(match[0]);
+  }
 
-  for (size_t i = 0; i < a.size(); ++i) {
-    ARROW_ASSIGN_OR_RAISE(auto match, a[i].FindOne(schema));
-    bool is_found = (b_paths.find(match[0]) != b_paths.end());
-    if (!is_found) {
-      result.push_back(a[i]);
-    }
+  for (auto ref : keys) {
+    ARROW_ASSIGN_OR_RAISE(auto match, ref.FindOne(schema));
+    payload_fields.erase(match[0]);
   }
 
+  std::vector<FieldRef> result;
```

Comment:
Minor nit: I know it was this way before but in general it's nice if we can avoid names like `result` and use something meaningful like `payload_refs`